File: Nim/Sources/Day02.nim
import strutils
import strformat
import sequtils
import AoCLib

const InputData = "Day02.txt"

type
  State = object
    Data: seq[seq[string]]

let s =new State
s.Data = toSeq((RawDataPath2021 & InputData).lines).mapIt(it.multiReplace(("  "," "))).mapIt(it.split(' '))

proc Part01() =

    var
        myX :int=0
        myY :int=0

    for mySeq in s.Data: 
        var myInt:int=myseq[1].parseInt
        case myseq[0]
            of "forward":
                myX += myInt
            of "up":
                myY -= myInt
            of "down" :
                myY += myInt
            else:
                discard
    var myresult:int=myY*myX
    echo fmt"The answer to Day 02 part 1 is 1507611 .  Found is {myResult}"


proc Part02() =
    
    var
        myX :int=0
        myY :int=0
        myAim:int=0

    for mySeq in s.Data: 
        var myInt:int=myseq[1].parseInt
        case myseq[0]
            of "forward":
                myX += myInt
                myY += (myAim*myInt)
            of "up":
                myAim -= myInt
            of "down" :
                myAim += myInt
            else:
                discard
    var myresult:int =myX*myY
    echo fmt"The answer to Day 02 part 2 is 1880593125.  Found is {myResult}"

proc Execute*()=
    
    Part01()
    Part02()

File: Cshp/AoC2021CSharp/Day02.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AoC2021CSharp
{
    internal class Day02
    {
        const string InputData = "Day02.txt";

        struct State
        {
           
            public List<string[]> Data { get; set; }
        }

        static State s = default;



        public static  void Execute()
        {
            Part02();
            Part01();
            

        }
        
        public static void Part2()
        {
            s.Data = new List<string[]>((IEnumerable<string[]>)File.ReadAllLines(Constants.AoC2021Data + InputData).Select(x => x.ToLower().Split(",").ToList()));

            long myX = default;
            long myY = default;
            long myAim = default;
            foreach (string[] myItem in s.Data)
            {
                long myMove = long.Parse(myItem[1]);

                switch (myItem[0])
                {
                    case "forward":
                        myX = ++myMove;
                        myY += (myAim * myMove);
                        break;

                    case "up":
                        myAim -= myMove;
                        break;

                    case "down":
                        myAim += myMove;
                        break;

                }
            }
            long myResult = myX * myY;
            Console.WriteLine("The answer to Day 02 part 2 is 1880593125. Found Is {0}", myResult);
        }


        public static void Part01()
        {
            s.Data = new List<string[]>((IEnumerable<string[]>)File.ReadAllLines(Constants.AoC2021Data + InputData).Select(x => x.ToLower().Split(",").ToList()));
            
            long myX = default;
            long myY = default;

            foreach (string[] myItem in s.Data)
            {
                long myMove = long.Parse(myItem[1]);
                switch (myItem[0])
                {
                    case "forward": 
                        myX += myMove; 
                        break;
                    
                    case "up":                     
                        myY -= myMove; 
                        break;
                    
                    case "down":                    
                        myY += myMove; 
                        break;   
                }
            }

            long myResult = myX * myY;

            Console.WriteLine("The answer to Day 02 part 1 is 1507611.  Found is {0}", myResult);
        }

       
    }
}
